File: Fixit/models/Bible.js
class Book {
    constructor(name, abbr, chapters, index) {
        this.name = name;
        this.abbr = abbr;
        this.chapters = chapters;
        this.index = index;
    }
}

class BIBLE {
    static  GENESIS              = new Book('Genesis',         'Gen',     50,    1);
    static  EXODUS               = new Book('Exodus',          'Exod',    40,    2);
    static  LEVITICUS            = new Book('Leviticus',       'Lev',     27,    3);
    static  NUMBERS              = new Book('Numbers',         'Num',     36,    4);
    static  DEUTERONOMY          = new Book('Deuteronomy',     'Deut',    34,    5);
    static  JOSHUA               = new Book('Joshua',          'Josh',    24,    6);
    static  JUDGES               = new Book('Judges',          'Judg',    21,    7);
    static  RUTH                 = new Book('Ruth',            'Ruth',     4,    8);
    static  FIRST_SAMUEL         = new Book('1 Samuel',        '1Sam',    31,    9);
    static  SECOND_SAMUEL        = new Book('2 Samuel',        '2Sam',    24,   10);
    static  FIRST_KINGS          = new Book('1 Kings',         '1Kgs',    22,   11);
    static  SECOND_KINGS         = new Book('2 Kings',         '2Kgs',    25,   12);
    static  FIRST_CHRONICLES     = new Book('1 Chronicles',    '1Chr',    29,   13);
    static  SECOND_CHRONICLES    = new Book('2 Chronicles',    '2Chr',    36,   14);
    static  EZRA                 = new Book('Ezra',            'Ezra',    10,   15);
    static  NEHEMIAH             = new Book('Nehemiah',        'Neh',     13,   16);
    static  ESTHER               = new Book('Esther',          'Esth',    10,   17);
    static  JOB                  = new Book('Job',             'Job',     42,   18);
    static  PSALMS               = new Book('Psalms',          'Ps',     150,   19);
    static  PROVERBS             = new Book('Proverbs',        'Prov',    31,   20);
    static  ECCLESIASTES         = new Book('Ecclesiastes',    'Eccl',    12,   21);
    static  SONG_OF_SOLOMON      = new Book('Song of Solomon', 'Song',     8,   22);
    static  ISAIAH               = new Book('Isaiah',          'Isa',     66,   23);
    static  JEREMIAH             = new Book('Jeremiah',        'Jer',     52,   24);
    static  LAMENTATIONS         = new Book('Lamentations',    'Lam',      5,   25);
    static  EZEKIEL              = new Book('Ezekiel',         'Ezek',    48,   26);
    static  DANIEL               = new Book('Daniel',          'Dan',     12,   27);
    static  HOSEA                = new Book('Hosea',           'Hos',     14,   28);
    static  JOEL                 = new Book('Joel',            'Joel',     3,   29);
    static  AMOS                 = new Book('Amos',            'Amos',     9,   30);
    static  OBADIAH              = new Book('Obadiah',         'Obad',     1,   31);
    static  JONAH                = new Book('Jonah',           'Jonah',    4,   32);
    static  MICAH                = new Book('Micah',           'Mic',      7,   33);
    static  NAHUM                = new Book('Nahum',           'Nah',      3,   34);
    static  HABAKKUK             = new Book('Habakkuk',        'Hab',      3,   35);
    static  ZEPHANIAH            = new Book('Zephaniah',       'Zeph',     3,   36);
    static  HAGGAI               = new Book('Haggai',          'Hag',      2,   37);
    static  ZECHARIAH            = new Book('Zechariah',       'Zech',    14,   38);
    static  MALACHI              = new Book('Malachi',         'Mal',      4,   39);
    static  MATTHEW              = new Book('Matthew',         'Matt',    28,   40);
    static  MARK                 = new Book('Mark',            'Mark',    16,   41);
    static  LUKE                 = new Book('Luke',            'Luke',    24,   42);
    static  JOHN                 = new Book('John',            'John',    21,   43);
    static  ACTS                 = new Book('Acts',            'Acts',    28,   44);
    static  ROMANS               = new Book('Romans',          'Rom',     16,   45);
    static  FIRST_CORINTHIANS    = new Book('1 Corinthians',   '1Cor',    16,   46);
    static  SECOND_CORINTHIANS   = new Book('2 Corinthians',   '2Cor',    13,   47);
    static  GALATIANS            = new Book('Galatians',       'Gal',      6,   48);
    static  EPHESIANS            = new Book('Ephesians',       'Eph',      6,   49);
    static  PHILIPPIANS          = new Book('Philippians',     'Phil',     4,   50);
    static  COLOSSIANS           = new Book('Colossians',      'Col',      4,   51);
    static  FIRST_THESSALONIANS  = new Book('1 Thessalonians', '1Thess',   5,   52);
    static  SECOND_THESSALONIANS = new Book('2 Thessalonians', '2Thess',   3,   53);
    static  FIRST_TIMOTHY        = new Book('1 Timothy',       '1Tim',     6,   54);
    static  SECOND_TIMOTHY       = new Book('2 Timothy',       '2Tim',     4,   55);
    static  TITUS                = new Book('Titus',           'Titus',    3,   56);
    static  PHILEMON             = new Book('Philemon',        'Phlm',     1,   57);
    static  HEBREWS              = new Book('Hebrews',         'Heb',     13,   58);
    static  JAMES                = new Book('James',           'Jas',      5,   59);
    static  FIRST_PETER          = new Book('1 Peter',         '1Pet',     5,   60);
    static  SECOND_PETER         = new Book('2 Peter',         '2Pet',     3,   61);
    static  FIRST_JOHN           = new Book('1 John',          '1John',    5,   62);
    static  SECOND_JOHN          = new Book('2 John',          '2John',    1,   63);
    static  THIRD_JOHN           = new Book('3 John',          '3John',    1,   64);
    static  JUDE                 = new Book('Jude',            'Jude',     1,   65);
    static  REVELATION           = new Book('Revelation',      'Rev',     22,   66);

    static getBook(name) {
        return Object.values(this).find(book => book.name.toLowerCase() === name.toLowerCase());
    }

    static getAbbr(name) {
        return Object.values(this).find(book => book.name.toLowerCase() === name.toLowerCase()).abbr;
    }
}

File: rust/speed_experiment/src/bible.rs
use std::collections::HashMap;
use std::sync::OnceLock;

pub struct ChapterPtr {
    pub book:Book,
    pub chapter:usize,
    pub index:usize,
    pub translation:Option<String>
}

#[derive(Clone, Copy)]
pub struct Book {
    pub name: &'static str,
    pub abbr: &'static str,
    pub index: u8,
    pub chapters: u8
}

pub struct BIBLE;
static CHAPTERS_MAP: OnceLock<HashMap<usize, ChapterPtr>> = OnceLock::new();

impl BIBLE {
    pub const GENESIS:Book               = Book { name: "Genesis",           abbr: "Gen",     index:  1, chapters:  50 };
    pub const EXODUS:Book                = Book { name: "Exodus",            abbr: "Exod",    index:  2, chapters:  40 };
    pub const LEVITICUS:Book             = Book { name: "Leviticus",         abbr: "Lev",     index:  3, chapters:  27 };
    pub const NUMBERS:Book               = Book { name: "Numbers",           abbr: "Num",     index:  4, chapters:  36 };
    pub const DEUTERONOMY:Book           = Book { name: "Deuteronomy",       abbr: "Deut",    index:  5, chapters:  34 };
    pub const JOSHUA:Book                = Book { name: "Joshua",            abbr: "Josh",    index:  6, chapters:  24 };
    pub const JUDGES:Book                = Book { name: "Judges",            abbr: "Judg",    index:  7, chapters:  21 };
    pub const RUTH:Book                  = Book { name: "Ruth",              abbr: "Ruth",    index:  8, chapters:   4 };
    pub const FIRST_SAMUEL:Book          = Book { name: "1 Samuel",          abbr: "1Sam",    index:  9, chapters:  31 };
    pub const SECOND_SAMUEL:Book         = Book { name: "2 Samuel",          abbr: "2Sam",    index: 10, chapters:  24 };
    pub const FIRST_KINGS:Book           = Book { name: "1 Kings",           abbr: "1Kgs",    index: 11, chapters:  22 };
    pub const SECOND_KINGS:Book          = Book { name: "2 Kings",           abbr: "2Kgs",    index: 12, chapters:  25 };
    pub const FIRST_CHRONICLES:Book      = Book { name: "1 Chronicles",      abbr: "1Chr",    index: 13, chapters:  29 };
    pub const SECOND_CHRONICLES:Book     = Book { name: "2 Chronicles",      abbr: "2Chr",    index: 14, chapters:  36 };
    pub const EZRA:Book                  = Book { name: "Ezra",              abbr: "Ezra",    index: 15, chapters:  10 };
    pub const NEHEMIAH:Book              = Book { name: "Nehemiah",          abbr: "Neh",     index: 16, chapters:  13 };
    pub const ESTHER:Book                = Book { name: "Esther",            abbr: "Esth",    index: 17, chapters:  10 };
    pub const JOB:Book                   = Book { name: "Job",               abbr: "Job",     index: 18, chapters:  42 };
    pub const PSALMS:Book                = Book { name: "Psalms",            abbr: "Ps",      index: 19, chapters: 150 };
    pub const PROVERBS:Book              = Book { name: "Proverbs",          abbr: "Prov",    index: 20, chapters:  31 };
    pub const ECCLESIASTES:Book          = Book { name: "Ecclesiastes",      abbr: "Eccl",    index: 21, chapters:  12 };
    pub const SONG_OF_SOLOMON:Book       = Book { name: "Song of Solomon",   abbr: "Song",    index: 22, chapters:   8 };
    pub const ISAIAH:Book                = Book { name: "Isaiah",            abbr: "Isa",     index: 23, chapters:  66 };
    pub const JEREMIAH:Book              = Book { name: "Jeremiah",          abbr: "Jer",     index: 24, chapters:  52 };
    pub const LAMENTATIONS:Book          = Book { name: "Lamentations",      abbr: "Lam",     index: 25, chapters:   5 };
    pub const EZEKIEL:Book               = Book { name: "Ezekiel",           abbr: "Ezek",    index: 26, chapters:  48 };
    pub const DANIEL:Book                = Book { name: "Daniel",            abbr: "Dan",     index: 27, chapters:  12 };
    pub const HOSEA:Book                 = Book { name: "Hosea",             abbr: "Hos",     index: 28, chapters:  14 };
    pub const JOEL:Book                  = Book { name: "Joel",              abbr: "Joel",    index: 29, chapters:   3 };
    pub const AMOS:Book                  = Book { name: "Amos",              abbr: "Amos",    index: 30, chapters:   9 };
    pub const OBADIAH:Book               = Book { name: "Obadiah",           abbr: "Obad",    index: 31, chapters:   1 };
    pub const JONAH:Book                 = Book { name: "Jonah",             abbr: "Jonah",   index: 32, chapters:   4 };
    pub const MICAH:Book                 = Book { name: "Micah",             abbr: "Mic",     index: 33, chapters:   7 };
    pub const NAHUM:Book                 = Book { name: "Nahum",             abbr: "Nah",     index: 34, chapters:   3 };
    pub const HABAKKUK:Book              = Book { name: "Habakkuk",          abbr: "Hab",     index: 35, chapters:   3 };
    pub const ZEPHANIAH:Book             = Book { name: "Zephaniah",         abbr: "Zeph",    index: 36, chapters:   3 };
    pub const HAGGAI:Book                = Book { name: "Haggai",            abbr: "Hag",     index: 37, chapters:   2 };
    pub const ZECHARIAH:Book             = Book { name: "Zechariah",         abbr: "Zech",    index: 38, chapters:  14 };
    pub const MALACHI:Book               = Book { name: "Malachi",           abbr: "Mal",     index: 39, chapters:   4 };
    pub const MATTHEW:Book               = Book { name: "Matthew",           abbr: "Matt",    index: 40, chapters:  28 };
    pub const MARK:Book                  = Book { name: "Mark",              abbr: "Mark",    index: 41, chapters:  16 };
    pub const LUKE:Book                  = Book { name: "Luke",              abbr: "Luke",    index: 42, chapters:  24 };
    pub const JOHN:Book                  = Book { name: "John",              abbr: "John",    index: 43, chapters:  21 };
    pub const ACTS:Book                  = Book { name: "Acts",              abbr: "Acts",    index: 44, chapters:  28 };
    pub const ROMANS:Book                = Book { name: "Romans",            abbr: "Rom",     index: 45, chapters:  16 };
    pub const FIRST_CORINTHIANS:Book     = Book { name: "1 Corinthians",     abbr: "1Cor",    index: 46, chapters:  16 };
    pub const SECOND_CORINTHIANS:Book    = Book { name: "2 Corinthians",     abbr: "2Cor",    index: 47, chapters:  13 };
    pub const GALATIANS:Book             = Book { name: "Galatians",         abbr: "Gal",     index: 48, chapters:   6 };
    pub const EPHESIANS:Book             = Book { name: "Ephesians",         abbr: "Eph",     index: 49, chapters:   6 };
    pub const PHILIPPIANS:Book           = Book { name: "Philippians",       abbr: "Phil",    index: 50, chapters:   4 };
    pub const COLOSSIANS:Book            = Book { name: "Colossians",        abbr: "Col",     index: 51, chapters:   4 };
    pub const FIRST_THESSALONIANS:Book   = Book { name: "1 Thessalonians",   abbr: "1Thess",  index: 52, chapters:   5 };
    pub const SECOND_THESSALONIANS:Book  = Book { name: "2 Thessalonians",   abbr: "2Thess",  index: 53, chapters:   3 };
    pub const FIRST_TIMOTHY:Book         = Book { name: "1 Timothy",         abbr: "1Tim",    index: 54, chapters:   6 };
    pub const SECOND_TIMOTHY:Book        = Book { name: "2 Timothy",         abbr: "2Tim",    index: 55, chapters:   4 };
    pub const TITUS:Book                 = Book { name: "Titus",             abbr: "Titus",   index: 56, chapters:   3 };
    pub const PHILEMON:Book              = Book { name: "Philemon",          abbr: "Phlm",    index: 57, chapters:   1 };
    pub const HEBREWS:Book               = Book { name: "Hebrews",           abbr: "Heb",     index: 58, chapters:  13 };
    pub const JAMES:Book                 = Book { name: "James",             abbr: "Jas",     index: 59, chapters:   5 };
    pub const FIRST_PETER:Book           = Book { name: "1 Peter",           abbr: "1Pet",    index: 60, chapters:   5 };
    pub const SECOND_PETER:Book          = Book { name: "2 Peter",           abbr: "2Pet",    index: 61, chapters:   3 };
    pub const FIRST_JOHN:Book            = Book { name: "1 John",            abbr: "1John",   index: 62, chapters:   5 };
    pub const SECOND_JOHN:Book           = Book { name: "2 John",            abbr: "2John",   index: 63, chapters:   1 };
    pub const THIRD_JOHN:Book            = Book { name: "3 John",            abbr: "3John",   index: 64, chapters:   1 };
    pub const JUDE:Book                  = Book { name: "Jude",              abbr: "Jude",    index: 65, chapters:   1 };
    pub const REVELATION:Book            = Book { name: "Revelation",        abbr: "Rev",     index: 66, chapters:  22 };

    pub const TOTAL_CHAPTERS:i16 = 1189;

    pub fn chapters_map(chapter_index: usize, translation: Option<&str>) -> Option<ChapterPtr> {
        let map = CHAPTERS_MAP.get_or_init(|| {
            let mut map = HashMap::new();
            for iter in Iterate_Bible_Chapters() {
                map.insert(
                    iter.i,
                    ChapterPtr::new(iter.book, iter.chapter, iter.i, None),
                );
            }
            map
        });

        let ptr: &ChapterPtr = map.get(&chapter_index)?;


        
        return Some(ChapterPtr {
            book: ptr.book.clone(),
            chapter: ptr.chapter,
            index: ptr.index,
            translation: translation.map(str::to_owned),
        });
    }
}


#[derive(Clone, Copy)]
pub struct ChapterIter {
    pub i: usize,
    pub book: Book,
    pub chapter: u8,
}
pub struct IterateBibleChapters {
    i: usize,
    books: std::slice::Iter<'static, Book>,
    current_book: Option<Book>,
    chapter: u8,
}
impl Iterator for IterateBibleChapters {
    type Item = ChapterIter;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            match self.current_book {
                Some(book) if self.chapter <= book.chapters => {
                    let item = ChapterIter {
                        i: self.i,
                        book,
                        chapter: self.chapter,
                    };
                    self.i += 1;
                    self.chapter += 1;
                    return Some(item);
                }

                _ => {
                    self.current_book = self.books.next().copied();
                    self.chapter = 1;

                    if self.current_book.is_none() {
                        return None;
                    }
                }
            }
        }
    }
}